src/NumericalAlgorithms/Integration/GslQuadAdaptive.hpp

```diff
@@ -104,157 +112,167 @@ class GslQuadAdaptiveImpl {
- * constructor. For details on the algorithm see the GSL documentation on
- * `gsl_integration`.
+ * All templates take upper bounds to the absolute and relative error;
+ * 'tolerance_abs' and 'tolerance_rel(default = 0.0)', respectively.
+ * For details on the algorithm see the GSL documentation on `gsl_integration`.
  *
  * Here is an example how to use this class:
  * For the function:
  * \snippet Test_GslQuad.cpp integrated_function
  * the integration should look like:
  * \snippet Test_GslQuad.cpp integration_example
  */
-
-enum class IntegralType {
-  StandardGaussKronrod,            // gsl_integration_qag()
-  IntegrableSingularitiesPresent,  // gsl_integration_qags()
-  IntegrableSingularitiesKnown,  // gsl_integration_qagp(), domain given by pts
-  InfiniteInterval,              // gsl_integration_qagi()
-  UpperBoundaryInfinite,         // gsl_integration_qagiu()
-  LowerBoundaryInfinite          // gsl_integration_qagil()
-};
-
 template <IntegralType TheIntegralType, typename... Args>
 class GslQuadAdaptive;
 
+/*!
+ * The template for "StandardGaussKronrod" uses the QAG algorithm to employ an
+ * adaptive Gauss-Kronrod n-points integration rule. Its function takes a
+ * 'lower_bound' and 'upper_bound' to the integration region, an upper bound for
+ * the absolute error 'tolerance_abs' and a 'key'. The latter is an index to the
+ * array [15, 21, 31, 41, 51, 61], where each element denodes how many function
+ * evaluations take place in each subinterval.
+ * The documentation mentions that a higher-order rule serves better for
+ * smooth functions, whereas a lower-order rule saves time for functions with
+ * local difficulties, such as discontinuities.
+ */
 template <typename... Args>
 class GslQuadAdaptive<IntegralType::StandardGaussKronrod, Args...>
     : public detail::GslQuadAdaptiveImpl<Args...> {
  public:
   using detail::GslQuadAdaptiveImpl<Args...>::GslQuadAdaptiveImpl;
-  // gsl_integration_qag(), applies Gauss-Kronrod rules with key as index
-  // to no_points = [15, 21, 31, 41, 51, 61]
   Result operator()(double lower_boundary, double upper_boundary,
-                    double tolerance_abs, int key) const noexcept {
+                    double tolerance_abs, int key,
+                    double tolerance_rel = 0.) const noexcept {
     Result result;
     int status = gsl_integration_qag(
-        &(this->integrand_), lower_boundary, upper_boundary, tolerance_abs, 0.,
-        this->max_intervals_, key, this->workspace_.get(), &result.value,
-        &result.error);
-    if (status != 0.) {
-      ::Parallel::printf("error: %s\n", gsl_strerror(status));
-    };
+        &(this->integrand_), lower_boundary, upper_boundary, tolerance_abs,
+        tolerance_rel, this->max_intervals_, key, this->workspace_.get(),
+        &result.value, &result.error);
     return result;
   }
 };
 
+/*!
+ * The template for "IntegrableSingularitiesPresent" uses the concentrates new,
```

Comment:
I think there's something wrong with this sentence :)

Reply:
Ups yeah ^^